### PR TITLE
[Enhancement] Add consistent hashing for hdfs backend selector

### PR DESCRIPTION
**Pull request by dirtysalt**
## What type of PR is this：
- [ ] bugfix
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 161 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/ConsistentHashRing.java|0|48|00.00%|[21, 30, 31, 32, 33, 36, 37, 38, 39, 45, 46, 47, 48, 49, 50, 51, 52, 53, 56, 57, 58, 59, 60, 63, 66, 67, 68, 69, 70, 73, 76, 77, 78, 79, 80, 82, 83, 84, 85, 87, 88, 91, 92, 93, 96, 97, 100, 102]|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|1|00.00%|[37]|
|:large_blue_circle:|com/starrocks/common/util/RendezvousHashRing.java|0|42|00.00%|[23, 24, 25, 26, 29, 32, 33, 34, 35, 36, 37, 38, 39, 42, 43, 44, 47, 48, 49, 50, 53, 56, 62, 63, 64, 65, 67, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 81, 83, 84, 85, 87]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|0|1|00.00%|[50]|
|:large_blue_circle:|com/starrocks/qe/HDFSBackendSelector.java|0|69|00.00%|[65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 80, 83, 84, 85, 87, 88, 89, 90, 91, 93, 94, 97, 98, 100, 116, 117, 120, 121, 124, 125, 126, 127, 128, 129, 130, 132, 133, 136, 139, 140, 141, 144, 147, 148, 168, 169, 170, 171, 173, 174, 175, 178, 179, 180, 181, 182, 184, 188, 196, 197, 198, 201, 202, 206, 215]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 88 / 166 (53.01%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|1|00.00%|[37]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|0|1|00.00%|[50]|
|:large_blue_circle:|com/starrocks/qe/HDFSBackendSelector.java|0|72|00.00%|[66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 81, 84, 85, 86, 88, 89, 91, 92, 93, 94, 96, 97, 100, 101, 103, 119, 120, 123, 124, 127, 128, 129, 130, 131, 132, 133, 135, 136, 139, 142, 143, 144, 147, 150, 151, 155, 156, 157, 158, 178, 179, 180, 181, 183, 184, 185, 188, 189, 190, 191, 192, 194, 198, 206, 209, 210, 214, 223]|
|:large_blue_circle:|com/starrocks/common/util/ConsistentHashRing.java|47|49|95.92%|[87, 94]|
|:large_blue_circle:|com/starrocks/common/util/RendezvousHashRing.java|41|43|95.35%|[60, 75]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 110 / 188 (58.51%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|1|00.00%|[37]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|0|1|00.00%|[50]|
|:large_blue_circle:|com/starrocks/qe/HDFSBackendSelector.java|0|72|00.00%|[66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 81, 84, 85, 86, 88, 89, 91, 92, 93, 94, 96, 97, 100, 101, 103, 119, 120, 123, 124, 127, 128, 129, 130, 131, 132, 133, 135, 136, 139, 142, 143, 144, 147, 150, 151, 155, 156, 157, 158, 178, 179, 180, 181, 183, 184, 185, 188, 189, 190, 191, 192, 194, 198, 206, 209, 210, 214, 223]|
|:large_blue_circle:|com/starrocks/common/util/ConsistentHashRing.java|47|49|95.92%|[87, 94]|
|:large_blue_circle:|com/starrocks/common/util/RendezvousHashRing.java|41|43|95.35%|[60, 75]|
|:large_blue_circle:|com/starrocks/common/util/PlainHashRing.java|22|22|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 114 / 192 (59.38%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|1|00.00%|[37]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|0|1|00.00%|[50]|
|:large_blue_circle:|com/starrocks/qe/HDFSBackendSelector.java|0|72|00.00%|[66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 81, 84, 85, 86, 88, 89, 91, 92, 93, 94, 96, 97, 100, 101, 103, 119, 120, 123, 124, 127, 128, 129, 130, 131, 132, 133, 135, 136, 139, 142, 143, 144, 147, 150, 151, 155, 156, 157, 158, 178, 179, 180, 181, 183, 184, 185, 188, 189, 190, 191, 192, 194, 198, 206, 209, 210, 214, 223]|
|:large_blue_circle:|com/starrocks/common/util/RendezvousHashRing.java|41|43|95.35%|[60, 75]|
|:large_blue_circle:|com/starrocks/common/util/ConsistentHashRing.java|51|53|96.23%|[97, 104]|
|:large_blue_circle:|com/starrocks/common/util/PlainHashRing.java|22|22|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 114 / 193 (59.07%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|1|00.00%|[37]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|0|1|00.00%|[50]|
|:large_blue_circle:|com/starrocks/qe/HDFSBackendSelector.java|0|73|00.00%|[60, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 81, 84, 85, 86, 88, 89, 91, 92, 93, 94, 96, 97, 100, 101, 103, 119, 120, 123, 124, 127, 128, 129, 130, 131, 132, 133, 135, 136, 139, 142, 143, 144, 147, 150, 151, 155, 156, 157, 158, 178, 179, 180, 181, 183, 184, 185, 188, 189, 190, 191, 192, 194, 198, 206, 209, 210, 214, 223]|
|:large_blue_circle:|com/starrocks/common/util/RendezvousHashRing.java|41|43|95.35%|[60, 75]|
|:large_blue_circle:|com/starrocks/common/util/ConsistentHashRing.java|51|53|96.23%|[97, 104]|
|:large_blue_circle:|com/starrocks/common/util/PlainHashRing.java|22|22|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 173 / 193 (89.64%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|1|00.00%|[37]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|0|1|00.00%|[50]|
|:large_blue_circle:|com/starrocks/qe/HDFSBackendSelector.java|59|73|80.82%|[72, 73, 74, 75, 76, 77, 78, 79, 86, 92, 93, 94, 124, 192]|
|:large_blue_circle:|com/starrocks/common/util/RendezvousHashRing.java|41|43|95.35%|[60, 75]|
|:large_blue_circle:|com/starrocks/common/util/ConsistentHashRing.java|51|53|96.23%|[97, 104]|
|:large_blue_circle:|com/starrocks/common/util/PlainHashRing.java|22|22|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/common/util/HashRing.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+package com.starrocks.common.util;
+
+import java.util.List;
+
+public interface HashRing<K, N> {
+    void addNode(N node);
+
+    void removeNode(N node);
+
+    List<N> get(K key, int distinctNumber);
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/common/util/HashRing.java:12]: Plz add some comments

File: fe/fe-core/src/main/java/com/starrocks/qe/HDFSBackendSelector.java [hunk 2]
```diff
@@ -60,97 +116,111 @@ public HDFSBackendSelector(ScanNode scanNode, List<TScanRangeLocations> location
...
+    private ComputeNode selectLeastScanBytesComputeNode(Collection<ComputeNode> backends) {
+        if (backends == null || backends.isEmpty()) {
+            return null;
+        }
+
+        ComputeNode node = null;
+        long minAssignedScanRanges = Long.MAX_VALUE;
+        for (ComputeNode backend : backends) {
+            long assignedScanRanges = assignedScansPerComputeNode.get(backend);
+            if (assignedScanRanges < minAssignedScanRanges) {
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/qe/HDFSBackendSelector.java:131]: Do we need to adjust assignment for balancing scan ranges in every single query.
First, the hash mechanism should assure the distribution is even.
Second, I'm afraid the adjustment will cause some scan range move from one node to another, causing cache miss.

Comment (dirtysalt) [reply fe/fe-core/src/main/java/com/starrocks/qe/HDFSBackendSelector.java:131]: > Do we need to adjust assignment for balancing scan ranges in every single query. 
Yes, we do

> First, the hash mechanism should assure the distribution is even. 
In theory yes,  but we also have to consider bad cases. If distirbution is not even, then there are some hotspots, we don't want to see that. 

> Second, I'm afraid the adjustment will cause some scan range move from one node to another, causing cache miss.

We select at most 3 different nodes, so cache data will be replicated at most 3 times.  Cache miss is not the only factor we should consider, we should also consider parallelism.  There is a balance between them.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 225 / 253 (88.93%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|1|00.00%|[37]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|0|1|00.00%|[50]|
|:large_blue_circle:|com/starrocks/qe/HDFSBackendSelector.java|59|73|80.82%|[72, 73, 74, 75, 76, 77, 78, 79, 86, 92, 93, 94, 124, 192]|
|:large_blue_circle:|com/starrocks/common/util/StaticPartitionedHashRing.java|50|57|87.72%|[80, 81, 82, 83, 84, 92, 99]|
|:large_blue_circle:|com/starrocks/common/util/RendezvousHashRing.java|43|46|93.48%|[60, 75, 85]|
|:large_blue_circle:|com/starrocks/common/util/ConsistentHashRing.java|51|53|96.23%|[97, 104]|
|:large_blue_circle:|com/starrocks/common/util/PlainHashRing.java|22|22|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 175 / 196 (89.29%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|1|00.00%|[37]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|0|1|00.00%|[50]|
|:large_blue_circle:|com/starrocks/qe/HDFSBackendSelector.java|59|73|80.82%|[72, 73, 74, 75, 76, 77, 78, 79, 86, 92, 93, 94, 124, 192]|
|:large_blue_circle:|com/starrocks/common/util/RendezvousHashRing.java|43|46|93.48%|[60, 75, 85]|
|:large_blue_circle:|com/starrocks/common/util/ConsistentHashRing.java|51|53|96.23%|[97, 104]|
|:large_blue_circle:|com/starrocks/common/util/PlainHashRing.java|22|22|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/qe/HDFSBackendSelector.java (hunk 2 above)
Comment (DorianZheng) [reply fe/fe-core/src/main/java/com/starrocks/qe/HDFSBackendSelector.java:131]: Maybe we could try to do that after all scan range assigned. It's hard to know whether it's uniformly distributed before the assignment finish.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 176 / 198 (88.89%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|1|00.00%|[37]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|0|1|00.00%|[50]|
|:large_blue_circle:|com/starrocks/qe/HDFSBackendSelector.java|59|73|80.82%|[72, 73, 74, 75, 76, 77, 78, 79, 86, 92, 93, 94, 124, 192]|
|:large_blue_circle:|com/starrocks/common/util/RendezvousHashRing.java|43|46|93.48%|[60, 75, 85]|
|:large_blue_circle:|com/starrocks/common/util/PlainHashRing.java|23|24|95.83%|[48]|
|:large_blue_circle:|com/starrocks/common/util/ConsistentHashRing.java|51|53|96.23%|[97, 104]|

File: fe/fe-core/src/main/java/com/starrocks/qe/HDFSBackendSelector.java (hunk 2 above)
Comment (dirtysalt) [reply fe/fe-core/src/main/java/com/starrocks/qe/HDFSBackendSelector.java:131]: I don't get it how it works, could you elaborate a little?

Let's say we have 4 scan ranges (S1,S2,S3, S4) and 4 hosts(H1 - H4).  For S1, we get H1-H3 as candidates, which host do we choose?

**Review by stephen-shelby — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/qe/HDFSBackendSelector.java (hunk 2 above)
Comment (dirtysalt) [reply fe/fe-core/src/main/java/com/starrocks/qe/HDFSBackendSelector.java:131]: I almost get what you mean. I'll write another version of it and you can review.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 192 / 214 (89.72%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|1|00.00%|[37]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|0|1|00.00%|[50]|
|:large_blue_circle:|com/starrocks/qe/HDFSBackendSelector.java|75|89|84.27%|[73, 74, 75, 76, 77, 78, 79, 80, 87, 93, 94, 95, 125, 215]|
|:large_blue_circle:|com/starrocks/common/util/RendezvousHashRing.java|43|46|93.48%|[60, 75, 85]|
|:large_blue_circle:|com/starrocks/common/util/PlainHashRing.java|23|24|95.83%|[48]|
|:large_blue_circle:|com/starrocks/common/util/ConsistentHashRing.java|51|53|96.23%|[97, 104]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 219 / 241 (90.87%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|1|00.00%|[37]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|0|1|00.00%|[50]|
|:large_blue_circle:|com/starrocks/qe/HDFSBackendSelector.java|102|116|87.93%|[74, 75, 76, 77, 78, 79, 80, 81, 88, 94, 95, 96, 126, 257]|
|:large_blue_circle:|com/starrocks/common/util/RendezvousHashRing.java|43|46|93.48%|[60, 75, 85]|
|:large_blue_circle:|com/starrocks/common/util/PlainHashRing.java|23|24|95.83%|[48]|
|:large_blue_circle:|com/starrocks/common/util/ConsistentHashRing.java|51|53|96.23%|[97, 104]|

**Review by Youngwb — APPROVED** [review]

**Review by stephen-shelby — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by dirtysalt** [comment]
run starrocks_admit_test

**Pull request merged**